File: src/main/java/space/maxkonkin/nasapicbot/client/YandexTranslateApiClient.java
package space.maxkonkin.nasapicbot.client;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.springframework.stereotype.Component;
import space.maxkonkin.nasapicbot.config.YandexTranslateApiConfig;
import space.maxkonkin.nasapicbot.model.LangCode;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.List;

@Component
public class YandexTranslateApiClient {
    private final YandexTranslateApiConfig config;

    private static final ObjectMapper mapper = new ObjectMapper();

    public YandexTranslateApiClient(YandexTranslateApiConfig config) {
        this.config = config;
    }

    public List<String> translate(List<String> inputTexts, LangCode langCode) throws IOException {
        final HttpPost httpPost = new HttpPost(config.getAPI_BASE_URI());
        httpPost.setHeader("Content-type", "application/json");
        httpPost.setHeader("Authorization", config.getAPI_TOKEN());

        final StringEntity entity = getStringEntity(inputTexts, langCode);
        httpPost.setEntity(entity);
        List<JsonNode> textNodes;
        try (CloseableHttpClient client = HttpClients.createDefault();
             CloseableHttpResponse response = client.execute(httpPost)) {
            JsonNode jsonNode = mapper.readTree(response.getEntity().getContent());
            textNodes = jsonNode.findValues("text");
        }

        return textNodes.stream().map(JsonNode::textValue).toList();
    }

    private StringEntity getStringEntity(List<String> inputTexts, LangCode langCode) throws UnsupportedEncodingException {
        StringBuilder commaSeparatedTexts = new StringBuilder();
        for (String inputText : inputTexts) {
            String filteredInputText = inputText.replace("\"","\\\"");
            commaSeparatedTexts.append("\"");
            commaSeparatedTexts.append(filteredInputText);
            commaSeparatedTexts.append("\",");
        }

        final String json = "{\n" +
                "    \"folderId\": \"" + config.getFOLDER_ID() + "\",\n" +
                "    \"texts\": [" + commaSeparatedTexts + "],\n" +
                "    \"targetLanguageCode\": \"" + langCode.getCode() + "\"\n" +
                "}";

        return new StringEntity(json);
    }
}

File: src/main/kotlin/space/maxkonkin/nasapicbot/client/YandexTranslateApiClient.kt
package space.maxkonkin.nasapicbot.client

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import space.maxkonkin.nasapicbot.config.YandexTranslateConfig
import space.maxkonkin.nasapicbot.model.LangCode

class YandexTranslateApiClient(
    private val config: YandexTranslateConfig,
    private val mapper: ObjectMapper,
    private val httpClient: OkHttpClient
) {
    fun translate(inputTexts: List<String>, langCode: LangCode): List<String> {
        val requestBody = getStringEntity(inputTexts, langCode).toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(config.apiBaseUri)
            .addHeader("Content-type", "application/json")
            .addHeader("Authorization", config.apiToken)
            .post(requestBody)
            .build()
        val textNodes: List<JsonNode>
        httpClient.newCall(request).execute().use { response ->
            val jsonNode = mapper.readTree(response.body?.string())
            textNodes = jsonNode.findValues("text")
        }

        return textNodes.map { node -> node.textValue() }.toList()
    }

    private fun getStringEntity(inputTexts: List<String>, langCode: LangCode): String {
        val commaSeparatedTexts = StringBuilder()
        for (inputText in inputTexts) {
            val filteredInputText = inputText.replace("\"", "\\\"")
            commaSeparatedTexts.append("\"")
            commaSeparatedTexts.append(filteredInputText)
            commaSeparatedTexts.append("\",")
        }
        val json =
            """{ "folderId": "${config.folderId}", "texts": [$commaSeparatedTexts], "targetLanguageCode": "${langCode.code}"}"""
        return json
    }
}